packages/react-data-grid/src/Viewport.tsx

```diff
@@ -66,233 +43,177 @@ type SharedGridProps<R> = Pick<GridProps<R>,
+      setIsScrolling(true);
+      resetScrollStateAfterDelay();
     }
 
-    const { height } = this.viewport.current.getBoundingClientRect();
-
-    if (height) {
-      const { scrollTop, scrollLeft } = this.state;
-      const { rowHeight, rowsCount } = this.props;
-      this.updateScroll({
-        scrollTop,
-        scrollLeft,
-        height,
-        rowHeight,
-        rowsCount
-      });
+    let previousScrollPosition: ScrollPosition | undefined;
+    if (scrollState !== null) {
+      previousScrollPosition = {
+        scrollLeft: scrollState.scrollLeft,
+        scrollTop: scrollState.scrollTop
+      };
     }
-  };
 
-  componentWillReceiveProps(nextProps: ViewportProps<R>) {
-    const { rowHeight, rowsCount } = nextProps;
-    if (this.props.rowHeight !== nextProps.rowHeight
-      || this.props.minHeight !== nextProps.minHeight) {
-      const { scrollTop, scrollLeft, height } = getGridState(nextProps);
-      this.updateScroll({
-        scrollTop,
-        scrollLeft,
-        height,
-        rowHeight,
-        rowsCount
-      });
-    } else if (this.props.columnMetrics.columns.length !== nextProps.columnMetrics.columns.length) {
-      this.setState(getGridState(nextProps));
-    } else if (this.props.rowsCount !== nextProps.rowsCount) {
-      const { scrollTop, scrollLeft, height } = this.state;
-      this.updateScroll({
-        scrollTop,
-        scrollLeft,
-        height,
+    const scrollDirection = getScrollDirection(previousScrollPosition, { scrollLeft, scrollTop });
+    const nextScrollState = {
+      scrollLeft,
+      scrollTop,
+      scrollDirection,
+      ...getVerticalRangeToRender({
+        height: canvasHeight,
         rowHeight,
-        rowsCount
-      });
-      // Added to fix the hiding of the bottom scrollbar when showing the filters.
-    } else if (this.props.rowOffsetHeight !== nextProps.rowOffsetHeight) {
-      const { scrollTop, scrollLeft } = this.state;
-      // The value of height can be positive or negative and will be added to the current height to cater for changes in the header height (due to the filer)
-      const height = this.state.height + this.props.rowOffsetHeight - nextProps.rowOffsetHeight;
-      this.updateScroll({
         scrollTop,
+        rowsCount,
+        scrollDirection,
+        overscanRowCount
+      }),
+      ...getHorizontalRangeToRender({
+        columnMetrics,
         scrollLeft,
-        height,
-        rowHeight,
-        rowsCount
-      });
-    }
-  }
+        viewportWidth: getDOMNodeOffsetWidth(),
+        scrollDirection,
+        overscanColumnCount
+      })
+    };
 
-  componentDidMount() {
-    window.addEventListener('resize', this.metricsUpdated);
-    this.metricsUpdated();
+    setScrollState(nextScrollState);
+    handleScroll(nextScrollState);
   }
 
-  componentWillUnmount() {
-    window.removeEventListener('resize', this.metricsUpdated);
-    this.clearScrollTimer();
-  }
+  useEffect(() => {
```

Comment:
You're setting a state in a `useEffect`, sounds like you want `useMemo` instead to replace both?